tools/wrapper/retrace.c

```diff
@@ -0,0 +1,93 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+
+/*
+ * print usage
+ */
+
+static void usage(const char *prog_name)
+{
+	fprintf(stderr, "Usage: %s -l <path of libretrace.so> <command line>\n",
+			prog_name);
+	exit(1);
+}
+
+/*
+ * fork process
+ */
+
+static void fork_process(int argc, char **argv)
+{
+	pid_t pid;
+	int dup_stdout, dup_stderr;
+
+	char **args;
+	int i;
+
+	/* set arguments */
+	args = (char **) malloc((argc - 1) * sizeof(char **));
+	if (!args) {
+		fprintf(stderr, "Out of memory!\n");
+		exit(1);
+	}
+
+	for (i = 0; i < argc - 3; i++)
+		args[i] = argv[i + 3];
+
+	/* open stdout */
+	dup2(STDOUT_FILENO, dup_stdout);
+	dup2(STDOUT_FILENO, dup_stderr);
+
+	/* fork new process */
+	pid = fork();
+	if (pid == 0) {
+		dup2(STDOUT_FILENO, dup_stdout);
+		dup2(STDERR_FILENO, dup_stderr);
+
+		/* set environment for preload */
+		setenv("LD_PRELOAD", argv[2], 1);
+		exit(execvp(argv[3], args));
+	} else if (pid < 0) {
+		fprintf(stderr, "Failed to fork new process(err:%d)\n", errno);
+		exit(1);
+	}
+
+	/* wait until the process has terminated */
+	do {
+		int w, status;
+
+		/* wait until process has terminated */
+		w = waitpid(pid, &status, -1);
+		if (w < 0) {
+			fprintf(stderr, "waitpid() error for command\n");
+			break;
+		}
+
+		/* set exit status */
+		if (WIFEXITED(status))
+			break;
+	} while (1);
```

Comment:
It will be more useful if `fork_process` returns the pid and `wait` happens elsewhere. If `fork_process` blocks, then you'll need another process/thread for additional tracees - remember the spawn utility?